README
// // $Header: /home/daffy/u0/vern/flex/RCS/README,v 2.9 93/02/06 21:08:38 vern Exp $

This is release 2.3 of flex - a full release.

The flex distribution consists of the following files:

	README		This message

	Makefile
	flexdef.h
	parse.y
	scan.l
	ccl.c
	dfa.c
	ecs.c		flex sources
	gen.c
	main.c
	misc.c
	nfa.c
	sym.c
	tblcmp.c
	yylex.c

	libmain.c	flex library (-lfl) source

	initscan.c	pre-flex'd version of scan.l

	flex.skel	skeleton for generated scanners

	flexdoc.1	full user documentation
	flex.1		reference documentation

	Changes		Differences between this release and the previous one

	COPYING		flex's copyright

	MISC/		a directory containing miscellaneous porting-related
			notes (for Atari, MS-DOS, Turbo-C, and VMS)


Decide where you want to keep flex.skel (suggestion:  /usr/local/lib),
but don't move it there yet.  Edit "Makefile" and change the definition
of SKELETON_FILE to reflect the full pathname of flex.skel.

Read the "Porting considerations" note in the Makefile and make
the necessary changes.

To make flex for the first time, use:

    make first_flex

which uses the pre-generated copy of the flex scanner (the scanner
itself is written using flex).

Assuming it builds successfully, you can test it using

    make test

The "diff" should not show any differences.

If you're feeling adventurous, issue "make bigtest" and be prepared
to wait a while.

Install flex using:

	make install


Please send problems and feedback to:

 	vern@ee.lbl.gov
 	ucbvax!ee.lbl.gov!vern

 	Computer Systems Engineering
 	46A/1123
 	Lawrence Berkeley Laboratory
 	1 Cyclotron Rd.
 	Berkeley, CA 94720


Changes
Changes between 2.3.8-7 and 2.3.8-8
	- flexskel.{cc,h} flex++.1.dman destructor does call YY_DELETE_BUFFER, and %define DESTRUCTOR_CODE. memory leakage avoided !
	- (same files)flexskel.{cc,h} flex++.1.dman %define CURRENT_BUFFER created to rename YY_CURRENT_BUFFER. there is no more yy_current_buffer member, but just the YY_CURRENT_BUFFER by default or YY_@_CURRENT_BUFFER.
Changes between 2.3.8-6 and 2.3.8-7
	- flexskel.cc flexskel.h flex++.1.dman: support iostream
	- samples show IOSTREAM calpability
Changes between 3.0.3 and 2.3.8-6
	- scan.l : add a header_skeleton_out() in rule ^"%%".* in
        order to include header skeleton section between section 1 and 2
Changes between 3.0.3 and 2.3.8-5
	- sent to CUG 
Changes between 3.0.3 and 2.3.8-4
	- main.c: change version name to (flex version-flex++ release), ie flex++-2.3.8-4
        - flexskel.cc : add #define YY_USE_CLASS if C++  line 39
        - main.cc :  line 876. accept the digits in #ifndef guard symbol name, instead of mapping them to underscore.
Changes between 3.0.2 and 3.0.3 
	- regenerated scan.c (and initscan.c) to use newest skeleton
	- Makefile:make test does not fail if diff fail, because it is always
         the case. The #line directives are always differents .
	- makefile: mail big test will fail on fast table, because there
	 is variable trailing context. don't worry
	- Notes.txt : creation of this file, with mail transaction
Version 3.0.2 sent to compiler.iecc.com
Changes between 3.0.1 and 3.0.2
	- flexskel.cc : remove include osfcn.h for dos
	- flexskel.cc : define _MSDOS if MSDOS defined or __MSDOS__
	- flexskel.cc : dquote remaining at end of line in debug code only. corrected.
	- main.c : mismatch between FILE* and char* when searching skeletons in getenv("INIT") on DOS.corrected.
	- main.c : mismatch that make .cpp suffix not to be replaced by .h in header. corrected.
	- flex++.1.dman : class declaration added in man page.
Changes between 3.0.0 and 3.0.1
	- use Patch #8 (21Feb93) of flex 2.3.8:
	- flex++.1.dman : remove stupid reference to %define PARSE_PARAM...
> Changes between 2.3 Patch #8 (21Feb93) and 2.3 Patch #7:
> 
> 	- Fixed bugs in dynamic memory allocation leading to grievous
> 	  fencepost problems when generating large scanners.
> 	- Fixed bug causing infinite loops on character classes with 8-bit
> 	  characters in them.
> 	- Fixed bug in matching repetitions with a lower bound of 0.
> 	- Fixed bug in scanning NUL characters using an "interactive" scanner.
> 	- Fixed bug in using yymore() at the end of a file.
> 	- Fixed bug in misrecognizing rules with variable trailing context.
> 	- Fixed bug compiling flex on Suns using gcc 2.
> 	- Fixed bug in not recognizing that input files with the character
> 	  ASCII 128 in them require the -8 flag.
> 	- Fixed bug that could cause an infinite loop writing out
> 	  error messages.
> 	- Fixed bug in not recognizing old-style lex % declarations if
> 	  followed by a tab instead of a space.
> 	- Fixed potential crash when flex terminated early (usually due
> 	  to a bad flag) and the -v flag had been given.
> 	- Added some missing declarations of void functions.
> 	- Changed to only use '\a' for __STDC__ compilers.
> 	- Updated mailing addresses.
> 
> 
Changes between 3.0.0 and 2.3.7 by coetmeur@icdc.fr
	- many things changed !!!
	- header can be generated
	- %header{ in section 1 and 2 
	- %define in section 1 
	- %name in section 1 
	- use 2 skeletons header and code flexskel.cc and flexskel.h
	- if no header generated, merge with code, else #include 
	- -g option change the name of #include 
	- -h option activate header generation and choose name 
	- -o option choose output name 
	- -h without name use lex.yy.h or the output (-o) appended/replaced with .h suffix. 
	- -H option change header skeleton 
	- skeleton support C++, and generate classes
	- many things can be changed by %define (names...)
	- compatibility with older version,  if no header generated.
	- C++ comment supported
	- #line automaticaly added in skeleton
	- coherent with bison++ version proposed by same author
Changes between 2.3 Patch #7 (28Mar91) and 2.3 Patch #6:

	- Fixed out-of-bounds array access that caused bad tables
	  to be produced on machines where the bad reference happened
	  to yield a 1.  This caused problems installing or running
	  flex on some Suns, in particular.


Changes between 2.3 Patch #6 (29Aug90) and 2.3 Patch #5:

	- Fixed a serious bug in yymore() which basically made it
	  completely broken.  Thanks goes to Jean Christophe of
	  the Nethack development team for finding the problem
	  and passing along the fix.


Changes between 2.3 Patch #5 (16Aug90) and 2.3 Patch #4:

	- An up-to-date version of initscan.c so "make test" will
	  work after applying the previous patches


Changes between 2.3 Patch #4 (14Aug90) and 2.3 Patch #3:

	- Fixed bug in hexadecimal escapes which allowed only digits,
	  not letters, in escapes
	- Fixed bug in previous "Changes" file!


Changes between 2.3 Patch #3 (03Aug90) and 2.3 Patch #2:

	- Correction to patch #2 for gcc compilation; thanks goes to
	  Paul Eggert for catching this.


Changes between 2.3 Patch #2 (02Aug90) and original 2.3 release:

	- Fixed (hopefully) headaches involving declaring malloc()
	  and free() for gcc, which defines __STDC__ but (often) doesn't
	  come with the standard include files such as <stdlib.h>.
	  Reordered #ifdef maze in the scanner skeleton in the hope of
	  getting the declarations right for cfront and g++, too.

	- Note that this patch supercedes patch #1 for release 2.3,
	  which was never announced but was available briefly for
	  anonymous ftp.


Changes between 2.3 (full) release of 28Jun90 and 2.2 (alpha) release:

    User-visible:

	- A lone <<EOF>> rule (that is, one which is not qualified with
	  a list of start conditions) now specifies the EOF action for
	  *all* start conditions which haven't already had <<EOF>> actions

testscan.c
#define FLEX_DEBUG
#define YY_CHAR unsigned char
# line 1 "/u/coetmeur/tools/flex++2/flexskel.cc"
/* A lexical scanner generated by flex */
/* scanner skeleton version:
 * $Header: /usr/fsys/odin/a/vern/flex/RCS/flex.skel,v 2.16 90/08/03 14:09:36 vern Exp $
 */
/* MODIFIED FOR C++ CLASS BY Alain Coetmeur: coetmeur(at)icdc.fr */
/* Note that (at) mean the 'at' symbol that I cannot write */
/* because it is expanded to the class name */
/* made at Informatique-CDC, Research&development department */
/* company from the Caisse Des Depots et Consignations */
/* institutional financial group (say 'Cat Doc Cad') */

/* theses symbols are added before this file */
/* #define YY_CHAR 'unsigned char' if 8bit or 'char' if 7bit */
/* #define FLEX_DEBUG if debug mode */
#define FLEX_SCANNER
/* cfront 1.2 defines "c_plusplus" instead of "__cplusplus" */
#ifdef c_plusplus
#ifndef __cplusplus
#define __cplusplus
#endif
#endif
#ifdef __cplusplus
#include <stdlib.h>
#define YY_USE_CONST
char *malloc();
int free();
int read();
#define YY_USE_PROTOS
#include <osfcn.h>
#else	/* ! __cplusplus */
#ifdef __STDC__
#ifdef __GNUC__
#include <stddef.h>
void *malloc( size_t );
void free( void* );
int read();
#else
#define YY_USE_CONST
#include <stdlib.h>
char *malloc();
int free();
int read();
#endif	/* __GNUC__ */
#define YY_USE_PROTOS
#endif	/* __STDC__ */
#endif	/* ! __cplusplus */
#ifdef __TURBOC__
#define YY_USE_CONST
#endif
#include <stdio.h>
/*********************************************/
/* COMPILER DEPENDENT   MACROS               */
/*********************************************/
/* use prototypes in function declarations */
/* the "const" storage-class-modifier is valid */
#ifndef YY_USE_CONST
#define const
#endif
/* use prototypes in function declarations */
#ifndef YY_PROTO
#ifdef YY_USE_PROTOS
#define YY_PROTO(proto) proto
#else
#define YY_PROTO(proto) ()
#endif
#endif


/*********************/
/* parameters        */

/* amount of stuff to slurp up with each read */
#ifndef YY_READ_BUF_SIZE
#define YY_READ_BUF_SIZE 8192
#endif
/* size of default input buffer */
#ifndef YY_BUF_SIZE
#define YY_BUF_SIZE (YY_READ_BUF_SIZE * 2) 
#endif

/***********************************/
/* to be redefined for application */

/* returned upon end-of-file */
#define YY_END_TOK 0
/* no semi-colon after return; correct usage is to write "yyterminate();" -
 * we don't want an extra ';' after the "return" because that will cause
 * some compilers to complain about unreachable statements.
 */
#define yyterminate() return ( YY_NULL )

/* code executed at the end of each rule */
#define YY_BREAK break;

/* #define YY_USER_ACTION */
/* #define YY_USER_INIT */


#ifndef YY_USE_CLASS
/* copy whatever the last rule matched to the standard output */
/* cast to (char *) is because for 8-bit chars, yytext is (unsigned char *) */
/* this used to be an fputs(), but since the string might contain NUL's,
 * we now use fwrite()
 */
#define ECHO (void) fwrite( (char *) yytext, yyleng, 1, yyout )

/* gets input and stuffs it into "buf".  number of characters read, or YY_NULL,
 * is returned in "result".
 */
#define YY_INPUT(buf,result,max_size) \
	if ( (result = read( fileno(yyin), (char *) buf, max_size )) < 0 ) \
	    YY_FATAL_ERROR( "read() in flex scanner failed" );

/* report a fatal error */

/* The funky do-while is used to turn this macro definition into
 * a single C statement (which needs a semi-colon terminator).
 * This avoids problems with code like:
 *
 * 	if ( something_happens )
 *		YY_FATAL_ERROR( "oops, the something happened" );
 *	else
 *		everything_okay();
 *
 * Prior to using the do-while the compiler would get upset at the
 * "else" because it interpreted the "if" statement as being all
 * done when it reached the ';' after the YY_FATAL_ERROR() call.
 */

#define YY_FATAL_ERROR(msg) \
	do \
		{ \
		(void) fputs( msg, stderr ); \
		(void) putc( '\n', stderr ); \
		exit( 1 ); \
		} \
	while ( 0 )

/* default yywrap function - always treat EOF as an EOF */
#define yywrap() 1


/* default declaration of generated scanner - a define so the user can
 * easily add parameters
 */
#define YY_DECL int yylex YY_PROTO(( void )) 
#else 
/* c++ */
#define ECHO (void) fwrite( (char *) yytext, yyleng, 1, yyout )
#define YY_INPUT(buf,result,max_size) \
	if ( yy_input((char *)buf, result,max_size) < 0 ) \
	    YY_FATAL_ERROR( "yy_input() in flex scanner failed" );

#define YY_FATAL_ERROR(msg) yy_fatal_error(msg)
#define yywrap() yy_wrap()

#endif
/***********************************/
/* not to be changed */
#define YY_NULL 0
#define YY_END_OF_BUFFER_CHAR 0
/* special action meaning "start processing a new file" */
#define YY_NEW_FILE \
	do \
		{ \
		yy_init_buffer( yy_current_buffer, yyin ); \
		yy_load_buffer_state(); \
		} \
	while ( 0 )
/* enter a start condition.  This macro really ought to take a parameter,
 * but we do it the disgusting crufty way forced on us by the ()-less
 * definition of BEGIN
 */
#define BEGIN yy_start = 1 + 2 *

/* action number for EOF rule of a given start state */
#define YY_STATE_EOF(state) (YY_END_OF_BUFFER + state + 1)



/* % section 1 definitions go here */ 
# line 1 "scan.l"
#define INITIAL 0
/* scan.l - scanner for flex input */
# line 5 "scan.l"
/*-
 * Copyright (c) 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Vern Paxson.
 * 
 * The United States Government has rights in this work pursuant
 * to contract no. DE-AC03-76SF00098 between the United States
 * Department of Energy and the University of California.
 *
 * Redistribution and use in source and binary forms are permitted provided
 * that: (1) source distributions retain this entire copyright notice and
 * comment, and (2) distributions including binaries display the following
 * acknowledgement:  ``This product includes software developed by the
 * University of California, Berkeley and its contributors'' in the
 * documentation or other materials provided with the distribution and in
 * all advertising materials mentioning features or use of this software.
 * Neither the name of the University nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

#ifndef lint
static char rcsid[] =
    "@(#) $Header: /usr/fsys/odin/a/vern/flex/RCS/scan.l,v 2.9 90/06/27 23:48:34 vern Exp $ (LBL)";
#endif

#undef yywrap

#include "flexdef.h"
#include "parse.h"

#define HEADER_ECHO fprintf(headerfile,"%s",yytext )
#define ACTION_ECHO fprintf( temp_action_file, "%s", yytext )
#define MARK_END_OF_PROLOG fprintf( temp_action_file, "%%%% end of prolog\n" );

#undef YY_DECL
#define YY_DECL \
	int flexscan()

#define RETURNCHAR \
	yylval = yytext[0]; \
	return ( CHAR );

#define RETURNNAME \
	(void) strcpy( nmstr, (char *) yytext ); \
	return ( NAME );

#define PUT_BACK_STRING(str, start) \
	for ( i = strlen( (char *) (str) ) - 1; i >= start; --i ) \
	    unput((str)[i])

#define CHECK_REJECT(str) \
	if ( all_upper( str ) ) \
	    reject = true;

#define CHECK_YYMORE(str) \
	if ( all_lower( str ) ) \
	    yymore_used = true;
